Homework 4: Sentiment Analysis - Task 3
----

Names & Sections
----
Names: Alec Condry (4120) and Shrihari Subramaniam (4120)

Task 3: Train a Logistic Regression Model (20 points)
----

Using `sklearn`'s implementation of `LogisticRegression`, conduct a similar analysis on the performance of a Logistic Regression classifier on the provided data set.

Using the `time` module, you'll compare and contrast how long it takes your home-grown BoW vectorizing function vs. `sklearn`'s `CountVectorizer`.


In [2]:
from sklearn.linear_model import LogisticRegression
# https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html
from sklearn.feature_extraction.text import CountVectorizer

from collections import Counter
import time
import sentiment_utils as sutils

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\hsubr\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [5]:
# define constants for the files we are using
TRAIN_FILE = "movie_reviews_train.txt"
DEV_FILE = "movie_reviews_dev.txt"

# load in your data and make sure you understand the format
# Do not print out too much so as to impede readability of your notebook
train_tups = sutils.generate_tuples_from_file(TRAIN_FILE)
dev_tups = sutils.generate_tuples_from_file(DEV_FILE)

# some variables you may want to use
BINARIZED = False
USE_COUNT_VECTORIZER = False

In [6]:
# Write the functions needed (here or in sentiment_utils.py) to create vectorized BoW representations
# of your data. We recommend starting with a multinomial BoW representation.
# Each training example should be represented as a sparse vector.

vocab = sutils.create_index(train_tups)

reviews_train = train_tups[0]
reviews_test = dev_tups[0]

y_train = train_tups[1]
y_test = dev_tups[1]

print(f'Vocab size: {len(vocab)}')

Vocab size: 30705


In [17]:
# how much time does it take to featurize the all data with your implementation?

start = time.time()

# YOUR CODE HERE
X_train_multi = sutils.featurize(vocab=vocab, data_to_be_featurized_X=reviews_train, binary=BINARIZED, verbose=False)
X_test_multi = sutils.featurize(vocab=vocab, data_to_be_featurized_X=reviews_test, binary=BINARIZED, verbose=False)

X_train_bi = sutils.featurize(vocab=vocab, data_to_be_featurized_X=reviews_train, binary=not BINARIZED, verbose=False)
X_test_bi = sutils.featurize(vocab=vocab, data_to_be_featurized_X=reviews_test, binary=not BINARIZED, verbose=False)

end = time.time()
print("That took:", end - start, "seconds")



That took: 55.17911720275879 seconds


In [6]:
# how much time does it take to featurize the all data with sklearn's CountVectorizer?
start = time.time()


# YOUR CODE HERE

end = time.time()
print("That took:", end - start, "seconds")

That took: 0.0 seconds


1. How big is your vocabulary using your vectorization function(s)? 30705
2. How big is your vocabulary using the `CountVectorizer`? __YOUR ANSWER HERE__

In [25]:
#  write any code you need analyze the relative sparsity of your vectorized representations of the data

# YOUR CODE HERE
percentages = []
for i in X_train_bi:
    counts = Counter(i)
    percentage = counts[0] / (counts[0] + counts[1])
    percentages.append(percentage * 100)


# Print out the average % of entries that are zeros in each vector in the vectorized training data
# YOUR CODE HERE
average = sum(percentages)/len(percentages)
print(f'Average % of entries that are zeros in each vector: {round(average, 2)}%')

Average % of entries that are zeros in each vector: 99.51%


In [8]:
# Using the provided dev set, evaluate your model with precision, recall, and f1 score as well as accuracy
# You may use nltk's implemented `precision`, `recall`, `f_measure`, and `accuracy` functions
# (make sure to look at the documentation for these functions!)
# you will be creating a similar graph for logistic regression and neural nets, so make sure
# you use functions wisely so that you do not have excessive repeated code
# write any helper functions you need in sentiment_utils.py (functions that you'll use in your other notebooks as well)


# create a graph of your classifier's performance on the dev set as a function of the amount of training data
# the x-axis should be the amount of training data (as a percentage of the total training data)
# the y-axis should be the performance of the classifier on the dev set
# the graph should have 4 lines, one for each of precision, recall, f1, and accuracy
# the graph should have a legend, title, and axis labels

# takes approx 30 sec on Felix's computer

Test the following 4 combinations to determine which has the best final f1 score for your Logistic Regression model:
- your vectorized features, multinomial: __enter your final f1 score here__
- CountVectorizer features, multinomial: __enter your final f1 score here__
- your vectorized features, binarized: __enter your final f1 score here__
- CountVectorizer features, binarized: __enter your final f1 score here__

Produce your graph(s) for the combination with the best final f1 score.




6120 REQUIRED
----

Find the top 100 most important features to your Logistic Regression classifier when using 100% of the training data. To access the weights of your model, you can access the `model.coef_` attribute. You'll want to use a `StandardScalar` preprocessor. This will help us deal with the fact that we expect counts of certain words to be higher (e.g. stop words).

To find the importance of a feature, calculate the absolute value of each weight in the model, then order your features according to the absolute values of these weights. The feature with the heighest absolute value weight has the most importance.

Use __your__ (not CountVectorizer) multinomial vectors for this analysis.

In [9]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler

In [10]:
# YOUR CODE HERE
# train a model on the scaled inputs
# This takes Felix's computer about 6.5 sec to run




In [11]:
# print out the top 20 most informative features according to this model


In [12]:
# re-evalaute your LR model with inputs that have been filtered to only use the top 500 most informative features


In [13]:
# create the same graph as before, but with the filtered inputs
